presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -218,42 +272,122 @@ private Aggregator(AccumulatorFactory accumulatorFactory, Step step)
-                aggregation.addInput(page);
+                accumulator.addInput(page);
             }
             else {
-                aggregation.addIntermediate(page.getBlock(intermediateChannel));
+                accumulator.addIntermediate(page.getBlock(intermediateChannel));
             }
         }
 
         public void evaluate(BlockBuilder blockBuilder)
         {
             if (step.isOutputPartial()) {
-                aggregation.evaluateIntermediate(blockBuilder);
+                accumulator.evaluateIntermediate(blockBuilder);
             }
             else {
-                aggregation.evaluateFinal(blockBuilder);
+                accumulator.evaluateFinal(blockBuilder);
+            }
+        }
+
+        public long getEstimatedSize()
+        {
+            return accumulator.getEstimatedSize();
+        }
+    }
+
+    private static class OrderingAccumulator implements Accumulator
+    {
+        private final Accumulator accumulator;
+        private final List<Integer> orderByChannels;
+        private final List<SortOrder> orderings;
+        private final int[] orderByOutputChannels;
+        private final PagesIndex pagesIndex;
+
+        private OrderingAccumulator(
+                Accumulator accumulator,
+                List<Type> orderBySourceTypes,
+                List<Integer> orderByChannels,
+                List<SortOrder> orderings,
+                PagesIndex.Factory pagesIndexFactory)
+        {
+            this.accumulator = requireNonNull(accumulator, "accumulator is null");
+            this.orderByOutputChannels = new int[orderBySourceTypes.size()];
+            for (int i = 0; i < orderBySourceTypes.size(); i++) {
+                orderByOutputChannels[i] = i;
             }
+            this.orderByChannels = ImmutableList.copyOf(orderByChannels);
+            this.orderings = orderings;
+            this.pagesIndex = pagesIndexFactory.newPagesIndex(orderBySourceTypes, 10_000);
         }
 
         public long getEstimatedSize()
         {
-            return aggregation.getEstimatedSize();
+            return pagesIndex.getEstimatedSize().toBytes() + accumulator.getEstimatedSize();
+        }
+
+        public Type getFinalType()
+        {
+            return accumulator.getFinalType();
+        }
+
+        public Type getIntermediateType()
+        {
+            return accumulator.getIntermediateType();
+        }
+
+        public void addInput(Page page)
+        {
+            pagesIndex.addPage(page);
+        }
+
+        public void addInput(WindowIndex index, List<Integer> channels, int startPosition, int endPosition)
+        {
+            throw new IllegalAccessError("Not supported");
+        }
+
+        public void addIntermediate(Block block)
+        {
+            throw new IllegalAccessError("Not supported");
+        }
+
+        public void evaluateIntermediate(BlockBuilder blockBuilder)
+        {
+            throw new IllegalAccessError("Not supported");
```

Comment:
here & above -- i don't think error is right here. `throw new UnsupportedOperationException();` should be enough, or `IllegalStateException("Not supported")`